frame/transaction-payment/src/payment.rs

```diff
@@ -0,0 +1,113 @@
+///! Traits and default implementation for paying transaction fees.
+use crate::{NegativeImbalanceOf, Trait};
+use codec::FullCodec;
+use frame_support::{
+	traits::{Currency, ExistenceRequirement, Get, Imbalance, OnUnbalanced, WithdrawReason},
+	unsigned::TransactionValidityError,
+};
+use sp_runtime::{
+	traits::{AtLeast32BitUnsigned, DispatchInfoOf, MaybeSerializeDeserialize, PostDispatchInfoOf},
+	transaction_validity::InvalidTransaction,
+};
+use sp_std::{fmt::Debug, marker::PhantomData};
+
+/// Handle withdrawing, refunding and depositing of transaction fees.
+pub trait OnChargeTransaction<T: Trait> {
+	type LiquidityInfo: Default;
+
+	/// Before the transaction is executed the payment of the transaction fees
+	/// need to be secured.
+	fn withdraw_fee(
+		who: &T::AccountId,
+		call: &T::Call,
+		dispatch_info: &DispatchInfoOf<T::Call>,
+		fee: T::Balance,
+		tip: T::Balance,
+	) -> Result<Self::LiquidityInfo, TransactionValidityError>;
+
+	/// After the transaction was executed and the correct fees where collected,
+	/// the resulting funds can be spend here.
+	fn deposit_fee(
+		who: &T::AccountId,
+		dispatch_info: &DispatchInfoOf<T::Call>,
+		post_info: &PostDispatchInfoOf<T::Call>,
+		fee: T::Balance,
+		tip: T::Balance,
+		liquidity_info: Self::LiquidityInfo,
+	) -> Result<(), TransactionValidityError>;
+}
+
+/// Implements the transaction payment for a module implementing the `Currency`
+/// trait (eg. the pallet_balances) using an unbalance handler (implementing
+/// `OnUnbalanced`).
+pub struct CurrencyAdapter<C, OU>(PhantomData<C>, PhantomData<OU>);
+
+/// Default implementation for a Currency and an OnUnbalanced handler.
+impl<T, C, OU, B> OnChargeTransaction<T> for CurrencyAdapter<C, OU>
+where
+	B: AtLeast32BitUnsigned + FullCodec + Copy + MaybeSerializeDeserialize + Debug + Default + Send + Sync,
+	T: Trait<Balance = B>,
+	T::TransactionByteFee: Get<B>,
+	C: Currency<<T as frame_system::Trait>::AccountId, Balance = B>,
+	C::PositiveImbalance: Imbalance<B, Opposite = C::NegativeImbalance>,
+	C::NegativeImbalance: Imbalance<B, Opposite = C::PositiveImbalance>,
+	OU: OnUnbalanced<NegativeImbalanceOf<C, T>>,
+{
+	type LiquidityInfo = Option<NegativeImbalanceOf<C, T>>;
+
+	/// Withdraw the predicted fee from the transaction origin.
+	fn withdraw_fee(
+		who: &T::AccountId,
+		_call: &T::Call,
+		_info: &DispatchInfoOf<T::Call>,
+		fee: B,
+		tip: B,
```

Comment:
I don't feel `tip` is something that `withdraw_fee` should know about. It is confusing in the sense that you immediately ask if `tip` is already part of `fee` here (it is). I know the parameter exists here because you need it in order to call `OU`. However, that is another thing I don't feel good about: Imho this shouldn't be hidden in a specific `OnChargeTransaction` implementation  but rather part of a trait. Otherwise anyone implementing `OnChargeTransaction` needs to repeat the splitting logic.

If the `OnUnbalanced` wasn't part of `CurrencyAdapter` there would be even the possibility to have a blanket `impl OnChargeTransaction for Currency`. 

 I am still fine with merging this the way it is, though.

Reply:
Yes it's confusing and unclear if the tip is included in the fee or not. I added some comments to clarify that (bfff36dfda57640bdf445a1367a0cf3ae9efd60b).

Regarding the splitting logic: Not every `OnChargeTransaction` has to call `OnUnbalanced`. Therefore not everyone implementing `OnChargeTransaction` would reimplement the splitting logic.

I also thought about removing the tip from the fee and always use the already splitted amount. Would that be better?

Reply:
`Yes it's confusing and unclear if the tip is included in the fee or not. I added some comments to clarify that (bfff36d).`

We need all the documentation on the trait so implementors know the expectations.

> I also thought about removing the tip from the fee and always use the already splitted amount (not adding the tip to the fee here). Would that be better?

No. Let's keep it included. It wouldn't make it less confusing just different.